lib/indexer/indexer.js

```diff
@@ -0,0 +1,690 @@
+    else
+      await this.saveGenesis();
+
+    // Bind to chain events.
+    this.bind();
+  }
+
+  /**
+   * Close the indexer, wait for the database to close.
+   * @returns {Promise}
+   */
+
+  async close() {
+    return this.db.close();
+  }
+
+  /**
+   * Ensure prefix directory (prefix/index).
+   * @returns {Promise}
+   */
+
+  async ensure() {
+    if (fs.unsupported)
+      return;
+
+    if (this.options.memory)
+      return;
+
+    await fs.mkdirp(this.options.prefix);
+  }
+
+  /**
+   * Verify network of index.
+   * @returns {Promise}
+   */
+
+  async verifyNetwork() {
+    let raw = await this.db.get(layout.O.encode());
+
+    if (!raw) {
+      raw = bio.write(4).writeU32(this.network.magic).render();
+      await this.db.put(layout.O.encode(), raw);
+      return;
+    }
+
+    const magic = bio.readU32(raw, 0);
+
+    if (magic !== this.network.magic)
+      throw new Error('Indexer: Network mismatch.');
+  }
+
+  /**
+   * A special case for indexing the genesis block. The genesis
+   * block coins are not spendable, however indexers can still index
+   * the block for historical and informational purposes.
+   * @private
+   * @returns {Promise}
+   */
+
+  async saveGenesis() {
+    this.start();
+
+    const block = Block.fromRaw(Buffer.from(this.network.genesisBlock, 'hex'));
+    const meta = new BlockMeta(block.hash(), 0);
+
+    await this.indexBlock(meta, block, new CoinView());
+    await this._setTip(meta);
+    await this.commit();
+
+    this.height = 0;
+  }
+
+  /**
+   * Bind to chain events.
+   * @private
+   */
+
+  bind() {
+    const listener = async (entry, block, view) => {
+      const meta = new BlockMeta(entry.hash, entry.height);
+
+      try {
+        await this.sync(meta, block, view);
+      } catch (e) {
+        this.emit('error', e);
+      }
+    };
+
+    this.chain.on('connect', listener);
+    this.chain.on('disconnect', listener);
+    this.chain.on('reset', listener);
+  }
+
+  /**
+   * Get a chain entry for the main chain only.
+   * @private
+   * @returns {Promise}
+   */
+
+  async getEntry(hash) {
+    const entry = await this.chain.getEntry(hash);
+
+    if (!entry)
+      return null;
+
+    if (!await this.chain.isMainChain(entry))
+      return null;
+
+    return entry;
+  }
+
+  /**
+   * Get a index block meta.
+   * @param {Hash} hash
+   * @returns {Promise}
+   */
+
+  async getBlockMeta(height) {
+    const data = await this.db.get(layout.h.encode(height));
+
+    if (!data)
+      return null;
+
+    return new BlockMeta(data, height);
+  }
+
+  /**
+   * Sync with the chain.
+   * @param {BlockMeta} meta
+   * @param {Block} block
+   * @param {CoinView} view
+   * @returns {Promise}
+   */
+
+  async sync(meta, block, view) {
+    if (this.syncing)
+      return;
+
+    this.syncing = true;
+
+    const connected = await this._syncBlock(meta, block, view);
+
+    if (connected) {
+      this.syncing = false;
+    } else {
+      (async () => {
```

Comment:
what's the purpose of wrapper function here?

Reply:
The rest of the behavior should return a Promise, and thus `await` for it to complete. However, this syncing should not and go ahead and continue. Thus, when an indexer is catching up with chain, it can do so in the background.

Reply:
yeah, it's hard to read in that format.
maybe move to 
```js
async syncChain() {
    try {
      await this._syncChain();
    } catch (e) {
      this.emit('error', e);
    } finally {
      this.syncing = false;
    }
}
```

Reply:
I don't think it's necessary, or would help to read. It's good to have `this.syncing = true` and each place it's set to false, within the same function.